### PR TITLE
US318359: Juned & Charlie: Added validation for cvv only functionality

### PR DESCRIPTION
**Pull request by juned-miah**
Created a new class that holds only cvv and then is able to perform validation based on cvv only

### CODE REVIEW

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/views/CVVLengthFilterTest.kt [hunk 1]
```diff
@@ -157,6 +199,18 @@ class CVVLengthFilterTest {
         assertNull(filtered)
     }
 
+    @Test
+    fun givenANullPanAndNoDefaultConfigurationThenShouldNotFilterBasedOnInput() {
```
Comment (ochalet-wp) [inline checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/views/CVVLengthFilterTest.kt]: Can we actually be in a case where we have no default configuration?

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verify(cardListener).onUpdate(cvvView, true)
+    }
+
+    @Test
+    fun givenCVVIsPartiallyAndCompleteInValidOnReValidationThenOnUpdateIsCalledWithValidTrue() {
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: done

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verify(cardListener).onUpdate(cvvView, true)
+    }
+
+    @Test
+    fun givenCVVIsPartiallyAndCompleteValidOnReValidationThenOnUpdateIsCalledWithValidTrue() {
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: I agree that tests should be independent and also that it is very hard to understand this without reading the implementation, however we have just simply kept things consistent with the AccessCheckoutCard even though I don't particularly agree with it.

It really comes down to priorities really, if we make client facing changes here, we would need to reflect the same for the other. I would rather do this as a refactor.

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        given(cardValidator.validateCVV(cvvValue, null)).willReturn(Pair(validationResult, null))
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView)
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        assertFalse { accessCheckoutCVV.isValid() }
+    }
+
+    @Test
+    fun givenCardValidatorIsSetThenACardShouldRevalidatePartialCVVFieldIfOnFocus() {
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: all names have been updated and should reflect the intention properly

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verify(cardListener).onUpdate(cvvView, true)
+    }
+
+    @Test
+    fun givenOnUpdateCVVIsCalledTheCVVIsPartiallyValidated() {
+        given(cvvView.hasFocus()).willReturn(false)
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: I would prefer the test to show this too, to be more specific on this assumption

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        accessCheckoutCVV.onUpdateCVV(cvvValue)
+
+        verify(cardListener).onUpdate(cvvView, true)
+    }
+
+    @Test
+    fun givenCardValidatorIsSetThenACardShouldRevalidateCompleteCVVFieldIfOnFocus() {
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: all names have been updated and should reflect the intention properly

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verify(cardListener).onUpdate(cvvView, true)
+        verifyZeroInteractions(cvvLengthFilter)
+    }
+
+
+    @Test
+    fun givenCardValidatorIsSetThenACardShouldValidateCompleteCVVFieldIfNotOnFocus() {
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: all names have been updated and should reflect the intention properly

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verify(cardListener).onUpdate(cvvView, true)
+
+    }
+
+    @Test
+    fun givenCardValidatorIsSetThenACardShouldValidateCompleteCVVFieldIfNotOnFocusAndPartialIsInvalid() {
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: all names have been updated and should reflect the intention properly

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verify(cardListener).onUpdate(cvvView, true)
+    }
+
+    @Test
+    fun givenCardValidatorIsSetThenACardShouldNotValidateCompleteCVVFieldIfNotOnFocusAndCompleteIsInvalid() {
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: all names have been updated and should reflect the intention properly

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/views/CVVLengthFilterTest.kt (hunk 1 above)
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/views/CVVLengthFilterTest.kt]: all names have been updated and should reflect the intention properly

**Pull request closed without merging**